lib/logstash/outputs/sqs.rb

```diff
@@ -1,136 +1,168 @@
-        )
+        raise LogStash::ConfigurationError, 'The batch size must be greater than 1'
       end
-      buffer_initialize(
-        :max_items => @batch_events,
-        :max_interval => @batch_timeout,
-        :logger => @logger
-      )
     end
 
     begin
-      @logger.debug("Connecting to AWS SQS queue '#{@queue}'...")
-      @sqs_queue = @sqs.queues.named(@queue)
-      @logger.info("Connected to AWS SQS queue '#{@queue}' successfully.")
-    rescue Exception => e
-      @logger.error("Unable to access SQS queue '#{@queue}': #{e.to_s}")
-    end # begin/rescue
-  end # def register
+      @logger.debug('Connecting to AWS SQS queue', :queue => @queue, :region => region)
+      @queue_url = @sqs.get_queue_url(:queue_name => @queue)[:queue_url]
+      @logger.info('Connected to AWS SQS queue successfully', :queue => @queue, :region => region)
+    rescue Aws::SQS::Errors::ServiceError => e
+      @logger.error('Failed to connect to Amazon SQS', :error => e)
+      raise LogStash::ConfigurationError, 'Verify the SQS queue name and your credentials'
+    end
+  end
 
   public
-  def receive(event)
+  def multi_receive_encoded(encoded_events)
     if @batch
-      buffer_receive(event.to_json)
-      return
+      multi_receive_encoded_batch(encoded_events)
+    else
+      multi_receive_encoded_single(encoded_events)
+    end
+  end
+
+  private
+  def multi_receive_encoded_batch(encoded_events)
+    bytes = 0
+    entries = []
+
+    # Split the events into multiple batches to ensure that a single batch does
+    # not exceed `@message_max_size`.
+    #
+    # TODO: We can probably do this more effectively. The current
+    # implementation simply ensures that the batch size is less than some
+    # threshold, but does not ensure that batch size itself is maximized.
+    # Maximizing the size of each batch would minimize the number of batch
+    # sends that need to be performed.
+    encoded_events.each_with_index.map do |event_and_data, index|
+      event, data = event_and_data
+
+      if data.bytesize > @message_max_size
+        @logger.warn('Message exceeds maximum length and will be dropped', :message_size => data.bytesize)
+        next
+      end
+
+      if (bytes + data.bytesize) > @message_max_size or entries.size >= @batch_events
+        @logger.debug("Publishing #{entries.size} messages to SQS")
+        @sqs.send_message_batch(:queue_url => @queue_url, :entries => entries)
+
+        bytes = 0
+        entries = []
+      end
+
+      bytes += data.bytesize
+      entries << { :id => index.to_s, :message_body => data }
     end
-    @sqs_queue.send_message(event.to_json)
-  end # def receive
 
-  # called from Stud::Buffer#buffer_flush when there are events to flush
-  def flush(events, close=false)
-    @sqs_queue.batch_send(events)
+    if entries.size > 0
+      @sqs.send_message_batch(:queue_url => @queue_url, :entries => entries)
+    end
   end
```

Comment:
```
 @sqs.send_message_batch(:queue_url => @queue_url, :entries => entries)
```
We could replace this logic into a small method that just takes the entries as an argument